solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol

```diff
@@ -0,0 +1,171 @@
+    ) external override ownerOnly {
+        totalPoolAmount[_poolToken] = totalPoolAmount[_poolToken].sub(_poolAmount);
+        totalReserveAmount[_poolToken][_reserveToken] = totalReserveAmount[_poolToken][_reserveToken].sub(_reserveAmount);
+        totalProviderAmount[_poolToken][_reserveToken][_provider] = totalProviderAmount[_poolToken][_reserveToken][_provider].sub(_reserveAmount);
+    }
+
+    /**
+     * @dev adds a pool to the list of pools of a liquidity provider
+     *
+     * @param _provider         liquidity provider address
+     * @param _poolToken        pool token address
+     */
+    function addProviderPool(
+        address _provider,
+        IDSToken _poolToken
+    ) external override ownerOnly returns (bool) {
+        return _providerPools[_provider].add(address(_poolToken));
+    }
+
+    /**
+     * @dev removes a pool from the list of pools of a liquidity provider
+     *
+     * @param _provider         liquidity provider address
+     * @param _poolToken        pool token address
+     */
+    function removeProviderPool(
+        address _provider,
+        IDSToken _poolToken
+    ) external override ownerOnly returns (bool) {
+        return _providerPools[_provider].remove(address(_poolToken));
+    }
+
+    /**
+     * @dev returns the list of pools of a liquidity provider
+     *
+     * @param _provider         liquidity provider address
+     * @return pool tokens
+     */
+    function providerPools(
+        address _provider
+    ) external override view returns (IDSToken[] memory) {
+        EnumerableSet.AddressSet storage set = _providerPools[_provider];
+        uint256 length = set.length();
+        IDSToken[] memory arr = new IDSToken[](length);
+        for (uint256 i = 0; i < length; i++) {
+            arr[i] = IDSToken(set.at(i));
+        }
+        return arr;
+    }
+
+    function seedPoolAmounts(
+        IDSToken[] calldata _tokens,
+        uint256[] calldata _amounts
+    ) external seederOnly {
+        uint256 length = _tokens.length;
+        for (uint256 i = 0; i < length; i++) {
+            totalPoolAmount[_tokens[i]] = _amounts[i];
+        }
+    }
+
+    function seedReserveAmounts(
+        IDSToken[] calldata _tokens,
+        IERC20Token[] calldata _reserves,
+        uint256[] calldata _amounts
+    ) external seederOnly {
+        uint256 length = _tokens.length;
+        for (uint256 i = 0; i < length; i++) {
+            totalReserveAmount[_tokens[i]][_reserves[i]] = _amounts[i];
+        }
+    }
+
+    function seedProviderAmounts(
+        IDSToken[] calldata _tokens,
+        IERC20Token[] calldata _reserves,
+        address[] calldata _providers,
+        uint256[] calldata _amounts
+    ) external seederOnly {
+        uint256 length = _tokens.length;
+        for (uint256 i = 0; i < length; i++) {
+            totalProviderAmount[_tokens[i]][_reserves[i]][_providers[i]] = _amounts[i];
+            _providerPools[_providers[i]].add(address(_tokens[i]));
```

Comment:
I think that the seeding of provider pools should be a separate function since we'd probably need seed provider amounts more than once, in which case handling provider pools, in the same call, will be redundant and cost additional gas

Reply:
The additional cost is neglectable, since function `EnumerableSet.add` checks before adding an existing item.